providers/implementations/keymgmt/ec_kmgmt.c

```diff
@@ -0,0 +1,592 @@
+    OSSL_PARAM_END
+};
+static const OSSL_PARAM *ec_types[] = {
+    NULL,                        /* Index 0 = none of them */
+    ec_domparameter_types,       /* Index 1 = domain parameter types */
+    ec_key_types,                /* Index 2 = key types */
+    ec_all_types                 /* Index 3 = 1 + 2 */
+};
+
+static
+const OSSL_PARAM *ec_imexport_types(int selection)
+{
+    int type_select = 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS) != 0)
+        type_select += 1;
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0)
+        type_select += 2;
+    return ec_types[type_select];
+}
+
+static
+const OSSL_PARAM *ec_import_types(int selection)
+{
+    return ec_imexport_types(selection);
+}
+
+static
+const OSSL_PARAM *ec_export_types(int selection)
+{
+    return ec_imexport_types(selection);
+}
+
+static
+int ec_get_params(void *key, OSSL_PARAM params[])
+{
+    EC_KEY *eck = key;
+    const EC_GROUP *ecg = NULL;
+    OSSL_PARAM *p;
+
+    ecg = EC_KEY_get0_group(eck);
+    if (ecg == NULL)
+        return 0;
+
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_MAX_SIZE)) != NULL
+        && !OSSL_PARAM_set_int(p, ECDSA_size(eck)))
+        return 0;
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_BITS)) != NULL
+        && !OSSL_PARAM_set_int(p, EC_GROUP_order_bits(ecg)))
+        return 0;
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_SECURITY_BITS)) != NULL) {
+        int ecbits, sec_bits;
+
+        ecbits = EC_GROUP_order_bits(ecg);
+
+        if (ecbits >= 512)
+            sec_bits = 256;
+        else if (ecbits >= 384)
+            sec_bits = 192;
+        else if (ecbits >= 384)
+            sec_bits = 192;
+        else if (ecbits >= 256)
+            sec_bits = 128;
+        else if (ecbits >= 224)
+            sec_bits = 112;
+        else if (ecbits >= 160)
+            sec_bits = 80;
+        else
+            sec_bits = ecbits / 2;
+
+        if (!OSSL_PARAM_set_int(p, sec_bits))
+            return 0;
+    }
+    return 1;
+}
+
+static const OSSL_PARAM ec_known_gettable_params[] = {
+    OSSL_PARAM_int(OSSL_PKEY_PARAM_BITS, NULL),
+    OSSL_PARAM_int(OSSL_PKEY_PARAM_SECURITY_BITS, NULL),
+    OSSL_PARAM_int(OSSL_PKEY_PARAM_MAX_SIZE, NULL),
+    OSSL_PARAM_int(OSSL_PKEY_PARAM_ECDH_COFACTOR_MODE, NULL),
+    OSSL_PARAM_END
+};
+
+static
+const OSSL_PARAM *ec_gettable_params(void)
+{
+    return ec_known_gettable_params;
+}
+
+static const OSSL_PARAM ec_known_settable_params[] = {
+    OSSL_PARAM_int(OSSL_PKEY_PARAM_ECDH_COFACTOR_MODE, NULL),
+    OSSL_PARAM_END
+};
+
+static
+const OSSL_PARAM *ec_settable_params(void)
+{
+    return ec_known_settable_params;
+}
+
+static
+int ec_set_params(void *key, const OSSL_PARAM params[])
+{
+    EC_KEY *eck = key;
+    const OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ECDH_COFACTOR_MODE);
+    if (p != NULL && !ec_set_param_ecdh_cofactor_mode(eck, p))
+        return 0;
+
+    return 1;
+}
```

Comment:
I'm trying to find code that makes this necessary, but am failing.  *You are already getting this value via import*, why is this needed?

Reply:
I would understand if you needed to set the cofactor mode because of some operation like SIGNATURE, KEYEXCH, whatever.  That code isn't here, though.

Reply:
I am working on an alt version of this that should address your concerns about the cofactor_mode flag.

In the meantime, my idea is that this "other param" should be settable from outside the provider, but it is true that currently we are missing an `EVP_PKEY_{set, get}_param()`, so currently the only way to do this is:

```C
EVP_PKEY *pkey = foo; // comes from somewhere
EC_KEY *eckey = EVP_PKEY_get0_EC_KEY(pkey); // reach inside the EVP for a legacy EC_KEY

if ((EC_KEY_get_flags(eckey) & EC_FLAG_COFACTOR_ECDH) == 0) {
    printf("Cofactor mode was not set, we force it!\n");
    EC_KEY_set_flags(eckey, EC_FLAG_COFACTOR_ECDH);
    /*
     * Use EC_KEY_clear_flags() to clear it
     */
} else {
    printf("Cofactor mode was already set, keep going!\n");
}
```

and rely on import/export mechanism.

Is this the only way that a provider should support? What about external providers that have gettable/settable parameters that do not map 1:1 with legacy DH, EC, DSA, RSA?

Reply:
Where would you have that code?  If it's part of an operation, why would you set a key flag rather than setting the cofactor flag for the operation in question?

Reply:
Uhm, because this particular parameter can be considered both a key parameter (as in "this is an ECCDH key, not a ECDH key" ) or as an operation parameter (for which we already have the code to set the mode on the operational EVP_PKEY_CTX)

Historically we have supported both modes, and that is why in 1.x we can set the flag either on EC_KEY or on EVP_PKEY_CTX. 

Also on EVP_PKEY_CTX_ctrl we have to support setting the mode to -1 with the meaning of "do ECDH or ECCDH depending on what is the default of the associated key".

Reply:
That did not answer my question.  *Where would the example code you showed above be?*

Reply:
I am feeling quite dumb at the moment!

Uhm, I guess my example code was supposed to be representative of what external applications might be doing at the moment if they are implementing something where being or not being an ECCDH key is seen as a key property rather than an operation property.

Reply:
OK.  The reason I'm being a bit insistent is that I was thinking this is something that's done when creating a key, which would typically be to create your `EC_KEY`, set that flag and assign it to an `EVP_PKEY` with 'EVP_PKEY_assign()`, or in the purely provided case, set up an `OSSL_PARAM` array including a cofactor element and create the `EVP_PKEY` with `EVP_PKEY_fromdata()`.

However, it sounds more like a need to tweak a key at any time post creation.  Yeahok, I guess I can deal with that.

Reply:
... and you're not dumb.  The whole cofactor mode thing is a confusing mess, so no wonder there are misunderstandings at times.